esphome/components/mqtt/mqtt_component.cpp

```diff
@@ -114,9 +114,17 @@ bool MQTTComponent::send_discovery_() {
         if (!unique_id.empty()) {
           root[MQTT_UNIQUE_ID] = unique_id;
         } else {
-          // default to almost-unique ID. It's a hack but the only way to get that
-          // gorgeous device registry view.
-          root[MQTT_UNIQUE_ID] = "ESP" + this->component_type() + this->get_default_object_id_();
+          const MQTTDiscoveryInfo &discovery_info = global_mqtt_client->get_discovery_info();
+          if (discovery_info.unique_id_generator == MQTT_MAC_ADDRESS_UNIQUE_ID_GENERATOR) {
+            char object_id_hash[9];
+            sprintf(object_id_hash, "%08x", fnv1_hash(this->get_default_object_id_()));
```

Comment:
This doesn't work, as `get_default_object_id()` already strips non-Latin characters. We should hash the raw name instead:
```suggestion
            sprintf(object_id_hash, "%08x", fnv1_hash(this->friendly_name()));
```